Add CEL for BackendTLSPolicy targetRefs

snorwin opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:

1. If this is your first time contributing to Gateway API, please read our
   developer guide (https://gateway-api.sigs.k8s.io/contributing/devguide/)
   and our community page (https://gateway-api.sigs.k8s.io/contributing/).
2. If this is your first time contributing to a Kubernetes project, please read
   our contributor guidelines:
   https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution
3. Please label this pull request according to what type of issue you are
   addressing, especially if this is a release targeted pull request. For
   reference on required PR/issue labels, read here:
   https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
4. If you want *faster* PR reviews, read how:
   https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it:
   https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
6. If this PR includes a new GEP please make sure you've followed the process
   outlined in our GEP overview, as this will help the community to ensure the
   best chance of positive outcomes for your proposal:
   https://gateway-api.sigs.k8s.io/geps/overview/#process
-->

**What type of PR is this?**
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind gep
/kind test

Optionally add one or more of the following kinds if applicable:
/kind deprecation
/kind failing-test
/kind flake
/kind regression
/area conformance
-->
/kind feature

**What this PR does / why we need it**:
This PR introduces validation to ensure that `targetRefs` on `BackendTLSPolicy` are unique. Without this validation, implementations may face challenges in properly resolving these references. A CEL is introduced, similar to the one used in the `HTTPRoute`, to facilitate this validation.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
None

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, please enter a release note below:
-->
```release-note
CEL validation for target references in BackendTLSPolicy
```

robscott commented on pkg/test/cel/backendtlspolicy_test.go:
```diff
@@ -25,29 +25,172 @@ import (
+			wantErrors: []string{},
+			targetRefs: []gatewayv1a2.LocalPolicyTargetReferenceWithSectionName{{
+				LocalPolicyTargetReference: gatewayv1a2.LocalPolicyTargetReference{
+					Group: gatewayv1a2.Group(corev1.GroupName),
+					Kind:  gatewayv1a2.Kind("Service"),
```
Would be good to add some test cases where the group or kind were the only parts that were different.

robscott reviewed (commented):
This is a great addition, thanks for the work on this @snorwin!

/cc @candita

robscott commented:
/ok-to-test

candita commented on apis/v1alpha3/backendtlspolicy_types.go:
```diff
@@ -71,6 +71,8 @@ type BackendTLSPolicySpec struct {
 	//
 	// +kubebuilder:validation:MinItems=1
 	// +kubebuilder:validation:MaxItems=16
+	// +kubebuilder:validation:XValidation:message="sectionName must be specified when targetRefs includes 2 or more references to the same target",rule="self.all(p1, self.all(p2, p1.group == p2.group && p1.kind == p2.kind && p1.name == p2.name ? ((!has(p1.sectionName) || p1.sectionName == '') == (!has(p2.sectionName) || p2.sectionName == '')) : true))"
```
Does this CEL apply itself appropriately when there are more than 2 `targetRefs`?  Maybe add another test case to verify that.

snorwin replied:
sure, I added some additional `targetRefs` in the test cases

robscott reviewed (approved):
Thanks @snorwin! Will defer final LGTM to @candita 

/approve

candita commented:
@snorwin It would be great if you could attend the next community meeting and do a small presentation on the background and solution, for the awareness of other implementers.  Also, a godoc update would be super helpful to all.

snorwin commented:
> @snorwin It would be great if you could attend the next community meeting and do a small presentation on the background and solution, for the awareness of other implementers. Also, a godoc update would be super helpful to all.

@candita  I added a description similar to the one for the `parentRefs` in `HTTPRoute`, but I've simplified it a bit for clarity.

I'll do my best to attend the community meeting, but the pre-xmas season is quite busy for me.

mlavacca reviewed (commented):
Thanks, @snorwin!

/approve

candita commented on apis/v1alpha3/backendtlspolicy_types.go:
```diff
@@ -65,6 +65,15 @@ type BackendTLSPolicySpec struct {
+	//   entries are distinct. In terms of fields, this means that the
+	//   multi-part key defined by `group`, `kind`, and `name` must
+	//   be unique across all targetRef entries in the BackendTLSPolicy.
+	// * They do not select different objects, each TargetRef that selects
+	//   the same object must set the sectionName to a different value.
```
nit:
```
	// * They do not select different objects, but each TargetRef that selects
	//   the same object must set the sectionName to a different value.
```

candita commented on apis/v1alpha3/backendtlspolicy_types.go:
```diff
@@ -65,12 +65,23 @@ type BackendTLSPolicySpec struct {
 	//
+	// TargetRefs must be _distinct_. This means either that:
+	//
+	// * They select different objects. If this is the case, then targetRef
+	//   entries are distinct. In terms of fields, this means that the
+	//   multi-part key defined by `group`, `kind`, and `name` must
+	//   be unique across all targetRef entries in the BackendTLSPolicy.
+	// * They do not select different objects, each TargetRef that selects
+	//   the same object must set the sectionName to a different value.
```
```suggestion
	// * They select different targets. If this is the case, then targetRef
	//   entries are distinct. In terms of fields, this means that the
	//   multi-part key defined by `group`, `kind`, and `name` must
	//   be unique across all targetRef entries in the BackendTLSPolicy.
	// * They select different sectionNames in the same target.
```

candita commented:
> > @snorwin It would be great if you could attend the next community meeting and do a small presentation on the background and solution, for the awareness of other implementers. Also, a godoc update would be super helpful to all.
> 
> @candita I added a description similar to the one for the `parentRefs` in `HTTPRoute`, but I've simplified it a bit for clarity.
> 
> I'll do my best to attend the community meeting, but the pre-xmas season is quite busy for me.

@snorwin I asked in the Slack channel.  I would prefer to make this more well-known now that we have 5 implementations and are aiming to move BackendTLSPolicy to standard.

robscott commented:
Thanks for the work on this @snorwin! I've added this to the agenda for our next meeting (Jan 6), and one of us can summarize the changes then (no worries if you can't make it). 

I think this is a strict improvement and unlikely to be problematic to implementations, but agree with @candita that we should make sure this is widely known in advance of the v1.3 release. Since this is on the agenda now for the next meeting, I'm fine with merging this at any point, but will defer to @candita on the timing.

candita commented on apis/v1alpha3/backendtlspolicy_types.go:
```diff
@@ -65,12 +65,22 @@ type BackendTLSPolicySpec struct {
 	// by default, but this default may change in the future to provide
 	// a more granular application of the policy.
 	//
+	// TargetRefs must be _distinct_. This means either that:
```
https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.SectionName shows `targetRef` section name would commonly be a port name on a Service, the expected type for `BackendTLSPolicy`.  For [LocalPolicyTargetReferenceWithSectionName](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1alpha2.LocalPolicyTargetReferenceWithSectionName), it says `sectionName` is optional and when not specified, refers to the entire resource.  

After some thought, I wonder if there is any use case where one `targetRef` needs to be associated to a specific `sectionName` (port name in the Service), and another `targetRef` doesn't care which port it targets on that Service.  In that case shouldn't it be okay to keep one `sectionName` blank, even though the other `sectionName` s would have to be specified in order to keep it distinct?

Also, I worry about the fact that `sectionName` is optional in `LocalPolicyTargetReferenceWithSectionName`.  If `LocalPolicyTargetReferenceWithSectionName` is used by another resource, will the same validation for distinct-ness (distinction?) apply, or do we foresee a need to require `sectionName` for distinction here and not when we use `LocalPolicyTargetReferenceWithSectionName` in another object?  And how confusing would the latter be?

candita replied:
cc @youngnick @robscott, since we talked about this today in the community meeting.

snorwin replied:
In the current version of the CEL, it is valid to keep a `sectionName` blank for one target (e.g., a `Service`) and specify `sectionName` for other targets. However, allowing the same target with and without `sectionName` and applying the most specific target could be confusing for both implementations and users.

Regarding the second concern that upcoming policies might use different validations for the `LocalPolicyTargetReferenceWithSectionName` type, we could either extend the description of the type or introduce a `CommonLocalPolicySpec `which could be reused, similar to the `CommonRouteSpec`.

candita replied:
The test "invalid because duplicate target refs with only one section name" implies it is not valid to keep a `sectionName`  blank for a duplicated targetRef.  Am I misunderstanding?

snorwin replied:
@candita no that's correct for duplicate target refs the `sectionName` must be specified (similar to the `HTTPRoute`).

candita replied:
@snorwin checking with @robscott about whether these are valid concerns.  If not, I'll remove the hold. Sorry for the delay.

robscott replied:
Thanks for the ping on this! I chatted a bit with @candita on Slack about this and agree with the changes in this PR. I think we can safely remove the hold here.

snorwin replied:
@candita and @robscott  thanks a lot 🙌🏻

candita commented:
/unhold

candita commented:
/lgtm
